### ISSUE TITLE
possible wrong CSE with unsafe casts

### ISSUE THREAD

**Issue by sandreenko** [dotnet/runtime#38920]
it is a follow-up from #38541

Try to compile and run the following test:
```
class TestStructs
{

    struct A
    {
        public int a;
    }

    // Make both structs with >4 fields to prevent struct promoting and force byref passing.
    struct CastFromStruct
    {
        public A primField;
        public int b;
        public int c;
        int d;
        int e;
    }

    struct CastToStruct
    {
        bool a;
        bool b;
        bool c;
        bool d;
        public int e; //b in CastFromStruct
    }

    struct CastToStruct2
    {
        bool a;
        bool b;
        bool c;
        bool d;
        public int e; //b in CastFromStruct
        public int f; //c in CastFromStruct
    }


    // our src local var has to be an implicit byref, otherwise we won't try to recover struct handle from it.
    [MethodImpl(MethodImplOptions.NoInlining)]
    static bool UnsafeCastFromAPrimitiveStructFieldToStruct(CastFromStruct impByRefStruct, int b)
    {
        CastToStruct to1 = new CastToStruct();
        // Force `IND struct(LCL_VAR   byref V00 arg0 Zero Fseq[primField])` where `primField` is a primitive type.
        to1 = Unsafe.As<A, CastToStruct>(ref impByRefStruct.primField);

        CastToStruct2 to2 = new CastToStruct2();
        to2 = Unsafe.As<A, CastToStruct2>(ref impByRefStruct.primField);

        Debug.Assert(to1.e == 1);
        Debug.Assert(to2.e == 1);
        Debug.Assert(to2.f == 2);

        return false;
    }
    public static int Main()
    {
        CastFromStruct s = new CastFromStruct();
        s.b = 1;
        s.c = 2;
        UnsafeCastFromAPrimitiveStructFieldToStruct(s, 1);
        return 100;
    }
}
```

it should run without issues, then try to run with `set complus_JitStressBiasedCSE=100`.
It should fail (x64 windows) with C# asserts because we will create a `primField` size CSE for `IND` in
```
               [000027] -A--G+------              *  ASG       struct (copy)
               [000025] D----+-N----              +--*  LCL_VAR   struct<CastToStruct2, 12> V03 loc1         
               [000024] n---G+------              \--*  IND       struct
               [000056] -----+-N----                 \--*  LCL_VAR   byref  V00 arg0          Zero Fseq[primField]
```
and when we read it as 8/12 bytes we won't see the original `CastFromStruct` fields.

I think it can't be reproduced without a stress mode, because for `IND->LCL_VAR byref` replacement cost is always bigger than profit but I am not 100% sure.

category:correctness
theme:cse
skill-level:expert
cost:medium

**Comment by sandreenko** [comment]
PTAL @briansull @dotnet/jit-contrib

**Comment by briansull** [comment]
I am not able to repro this, no CSE is created and the tree looks like this before/fter morph:

```
fgMorphTree BB01, STMT00003 (before)
               [000027] -ACXG-------              *  ASG       struct (copy)
               [000025] D------N----              +--*  LCL_VAR   struct<CastToStruct2, 12> V04 tmp1         
               [000018] --CXG-------              \--*  OBJ       struct<CastToStruct2, 12>
               [000017] --C-G-------                 \--*  CALL      byref  System.Runtime.CompilerServices.Unsafe.As
               [000016] ----G------- arg0               \--*  ADDR      byref 
               [000015] ----G--N----                       \--*  FIELD     struct primField
               [000014] ------------                          \--*  ADDR      byref 
               [000013] -------N----                             \--*  LCL_VAR   struct V00 arg0    

fgMorphTree BB01, STMT00003 (after)
               [000027] -ACXG+------              *  ASG       struct (copy)
               [000025] D----+-N----              +--*  LCL_VAR   struct<CastToStruct2, 12> V04 tmp1         
               [000018] --CXG+------              \--*  IND       struct
               [000017] --CXG+------                 \--*  CALL      byref  System.Runtime.CompilerServices.Unsafe.As
               [000013] -----+-N---- arg0 in rcx        \--*  LCL_VAR   byref  V00 arg0          Zero Fseq[primField]
```

**Comment by sandreenko** [comment]
Fixed by #40164
